File: android-mma-recomp/app/src/main/java/com/example/mmarecomp/viewmodel/ProfileViewModel.kt
package com.example.mmarecomp.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mmarecomp.data.ProfileRepository
import com.example.mmarecomp.model.Phase
import com.example.mmarecomp.model.Profile
import com.example.mmarecomp.model.ProfileUpdate
import com.example.mmarecomp.ui.components.ErrorOperation
import com.example.mmarecomp.ui.components.ScreenError
import com.example.mmarecomp.util.rethrowCancellation
import kotlinx.coroutines.launch

class ProfileViewModel(
    private val userId: String,
    private val repository: ProfileRepository = ProfileRepository(),
) : ViewModel() {
    var profile by mutableStateOf<Profile?>(null)
        private set
    var poidsObjectifKg by mutableStateOf("")
    var bfObjectifPct by mutableStateOf("")
    var phase by mutableStateOf(Phase.Ete)
    var isSaving by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set
    var errorOperation by mutableStateOf(ErrorOperation.LOAD)
        private set

    val screenError: ScreenError?
        get() = errorMessage?.let { ScreenError(it, errorOperation) }

    private var pendingSave: (() -> Unit)? = null

    private fun reportError(message: String, operation: ErrorOperation) {
        errorMessage = message
        errorOperation = operation
    }

    fun load() {
        viewModelScope.launch {
            try {
                val fetched = repository.fetch(userId)
                profile = fetched
                poidsObjectifKg = fetched.poidsObjectifKg.toString()
                bfObjectifPct = fetched.bfObjectifPct.toString()
                phase = fetched.phase
            } catch (e: Throwable) {
                rethrowCancellation(e)
                when (e) {
                    is java.io.IOException -> {
                reportError("Pas de connexion internet — réessaie dès que le réseau revient.", ErrorOperation.LOAD)
                    }
                    else -> {
                reportError("Impossible de charger le profil.", ErrorOperation.LOAD)
                    }
                }
            }
        }
    }

    fun save(onSaved: (Phase) -> Unit) {
        isSaving = true
        pendingSave = { save(onSaved) }
        viewModelScope.launch {
            val patch = ProfileUpdate(
                poidsObjectifKg = poidsObjectifKg.replace(",", ".").toDoubleOrNull(),
                bfObjectifPct = bfObjectifPct.replace(",", ".").toDoubleOrNull(),
                phase = phase,
            )
            try {
                repository.update(userId, patch)
                onSaved(phase)
                pendingSave = null
            } catch (e: Throwable) {
                rethrowCancellation(e)
                reportError("Impossible d'enregistrer le profil.", ErrorOperation.SAVE)
            } finally {
                isSaving = false
            }
        }
    }

    fun retrySave() {
        pendingSave?.invoke()
    }
}

File: ios-mma-recomp/MMARecomp/ViewModels/ProfileViewModel.swift
import Foundation

@MainActor
final class ProfileViewModel: ObservableObject {
    @Published var profile: Profile?
    @Published var poidsObjectifKg: String = ""
    @Published var bfObjectifPct: String = ""
    @Published var phase: Phase = .ete
    @Published var coachNotes: String = ""
    @Published var isSaving = false
    @Published var errorMessage: String?

    private let service = ProfileService()
    private let userId: UUID

    init(userId: UUID) {
        self.userId = userId
    }

    func load() async {
        do {
            let fetched = try await service.fetch(userId: userId)
            profile = fetched
            poidsObjectifKg = String(fetched.poidsObjectifKg)
            bfObjectifPct = String(fetched.bfObjectifPct)
            phase = fetched.phase
            coachNotes = fetched.coachNotes ?? ""
        } catch {
            errorMessage = "Impossible de charger le profil."
        }
    }

    func save() async {
        isSaving = true
        defer { isSaving = false }
        let patch = ProfileUpdate(
            poidsObjectifKg: Double(poidsObjectifKg.replacingOccurrences(of: ",", with: ".")),
            bfObjectifPct: Double(bfObjectifPct.replacingOccurrences(of: ",", with: ".")),
            phase: phase,
            coachNotes: coachNotes.isEmpty ? nil : coachNotes
        )
        do {
            try await service.update(userId: userId, patch: patch)
        } catch {
            errorMessage = "Impossible d'enregistrer le profil."
        }
    }
}
